Add ability to suppress --trace for certain error types

WilliamMcCumstie opened the pull request:
In some cases we do not want the `--trace` to be printed if there is an error. This is to allow errors to be easily distinguished between `ruby/developer` and `user` errors. The `--trace` option will still display the backtrace regardless. The only change is the user will not always be prompt to use it.

The implementation is fairly straight forward, an array of error types to be suppressed can be configured using `suppress_trace_classes`. This can take multiple error class inputs to be suppressed.

Any errors that inherit from a suppressed class will also be suppressed. This is to make it easier to suppress entire families of classes.

NOTE: The are some failing test that are unrelated to this PR. I have created a card for them
https://trello.com/c/yNAvzwVC/246-fix-commander-errors-due-to-erb-templating

WilliamMcCumstie committed:
Add method to suppress the --trace message

There are certain errors in which the "--trace" message should not be
displayed to the user. This is to distinguish between user errors (in
which the backtrace is not useful) and all other errors (in which it
is).

This commit updates the spec and adds the method

NOTE: If the "--trace" option is included, the backtrace will still
be displayed. This method only concerns whether the option is displayed
to the user.

WilliamMcCumstie committed:
Implement the suppression of --trace

The method has changed name slightly to be more descriptive, but
otherwise an array of classes are stored internally.

If an error class is raised that is included in the array, then --trace
is not displayed.

WilliamMcCumstie committed:
Suppress --trace inherited class as well

This way entire families can be suppressed without having to manually
added them all.

WilliamMcCumstie committed:
Bump the minor version number

bobwhitelock commented on lib/commander/runner.rb:
```diff
@@ -91,6 +91,18 @@ def version
+    # Suppresses the --trace message for these classes
+    # However --trace will still display the backtrace if included
+
+    def suppress_trace_classes(*classes)
```
Is there a need for this to take multiple classes? I think it would be simpler and clearer if there is just the option to have `Commander` suppress showing `--trace` for a single error class and its descendants.

This would also make things clearer in Metalware, as we would know there's only a single error class (with descendants) which this error message will not be shown for, and everything else it will be.

WilliamMcCumstie replied:
Fixed

bobwhitelock commented on lib/commander/version.rb:
```diff
@@ -1,3 +1,3 @@
 module Commander
-  VERSION = '4.5.3'.freeze
+  VERSION = '4.6.0'.freeze
```
I think independently bumping the Commander version number for our fork, as if this was the main Commander repo, is only likely to confuse us in future - maybe calling this `4.6.0-alces` would be clearer?

WilliamMcCumstie replied:
I have updated to be `4.4.3-alces0`. Version `4.4.3` was the point in which we forked the project. It will be useful to maintain as a reference to where we are compared to the original project.

bobwhitelock replied:
Good idea! :+1:

bobwhitelock reviewed (commented):
Mostly looks reasonable, just a couple of comments for you to consider. Also regarding that Trello card you've added, I don't think it makes sense to hide issues with our Commander fork within our Trello board for Metalware, as they aren't Metalware-specific and this repo may be used by other projects in future. I think it would be better to just log such issues in the issues for this repo.

WilliamMcCumstie commented:
Concerning the trello card.  As this is a fork of the original project, we can't log issues against it. So for the time being, I've included it in metalware.

We may want to consider making this our own repo as opposed to a fork. That way we can log issues.

WilliamMcCumstie committed:
Revert version numbering to match forked project

To prevent confusion in version numbering, the version has been reverted
to what it was originally. At the time this project was forked, the
version number was 4.4.3.

To distinguish between the original project and our updates, an
additional "alces" flag has been added.

For this reason the version number is "4.4.3-alces0". This way
the alces numbering can be bumped in the future.

NOTE: The previous version numbering will be maintained as older
versions of metalware depend on it.

https://github.com/alces-software/commander/pull/2#discussion_r168198886

bobwhitelock commented:
@WilliamMcCumstie I may be wrong but I believe you can turn issues on/off in the settings, they're just off by default for forks. I can't do this as I didn't create this repo.

WilliamMcCumstie commented:
TIL forks can have issues

WilliamMcCumstie committed:
Rename suppress_trace_classes to suppress_trace_class

This method will be adjusted to suppress only a single class

WilliamMcCumstie committed:
Only support a suppressing a single error family

Instead of allowing an array of classes to suppress, only allow one.
Atm this is sufficient functionality for our purposes.

Inherited classes are still suppressed

https://github.com/alces-software/commander/pull/2#discussion_r168197894

bobwhitelock commented:
LGTM
